### ISSUE TITLE
Feature Request: Ability to Import Sessions/URLs/csv

### ISSUE THREAD

**Issue by Aezaeo** [ReDEnergy/SessionSync#32]
I recently started to reuse Firefox after the Quantum updates so I'm finding myself trying to import sessions from Chrome's "Session Buddy". SessionSync is the closest I've found to "Session Buddy" in terms of layout and capabilities. There's a few things I'd like to see implemented but this is about  SessionSync's ability to Import. Is it possible to add an import function to allow imports of urls /sessions, or the abilty to make a backup of the Sessions you have created with SessionSync in case I want to reformat my pc or something along those lines. And I am aware of what syncing is, it's just sometimes fool proof to have a local backup in case things go wrong.

I've tried "Tab Session Manager" for Firefox to achieve the import function but it only allows URLs and doesn't properly read the .json that Chrome's "Session Buddy" creates. And I rather have a singular Add-on that does it all, rather than having multiple add-ons to make things work.

At the current time I'm having to use "Tab Session Manager" to Import "urls" from "Session Buddy" into Firefox, then Opening all the tabs in a Window to then create a session with "SessionSync". Doesn't seem like much but it's fairly tedious seeing as "Session Buddy" was extremely easy to use much like "SessionSync" and I created a lot of different categorized sessions with "Session Buddy".

Extra Feature Request; Another thing that "Session Buddy" had was the ability to sort urls alphabetically, I really enjoyed this feature to sort through links easier. I like that you have implemented a drag and drop feature to sort links, "Session Buddy" didn't have this. But when going though a session with a more than like 15 links can be tedious.

Thanks in advance and would love to hear what you think of these requests.

**Comment by ReDEnergy** [comment]
The main reason why I chose to have sessions represented as bookmarks was the fact that those bookmarks will be there no matter what will you do in the feature (using the same addon or something else or even move completely to a new browser) So... in that regard, the addon does not offer the ability to export the sessions to a backup file but you can do that by using bookmark export functionality from Firefox as a JSON file or HTML file (Bookmark Manager). 
To answer some of your questions:
1. If you plan to format your PC. Just make a backup of the bookmarks, and then import them again. Then the addon will detect them automatically.
  All your sessions can be found under **BookmarkMenu->SessionSync**. They will be there no matter if the addon is installed or not
2. You're not the only one who asked for an import feature so most probably I will try to offer a solution for that because right now the browser only knows how to export all the bookmarks, not just some of them. Exporting is easy... importing is a little bit harder with the current API but for that purpose native import can be used with success.
3. All bookmarks that you move into the **BookmarkMenu->SessionSync** will be visible by the SessionSync addon. So as long as you can import some bookmarks and move them there everything will be visible.
4. I do plan to add bookmarks sorting similar to the available options for sessions (by date, by name, by position and why not... by URL)

**Comment by x11x** [comment]
Hi,
I realise that this is not a solution to the problem of implementing the import/export feature on top of the bookmarks storage.
But I would just like to recommend supporting the simple, plain text format that SessionBuddy uses for import and export. Its quite intuitive and user friendly, using indents, supporting custom titles.
An example:
```
Window Title

   My example tab title
   http://www.example.com

   Another Tab
   http://www.example.com
```
The good thing about it is it supports up to 3 levels of hierarchy (Session, Window and Tab), but these levels are optional. You can just have a simple list of URLs with no indenting. You can also omit the titles so a simple list of URLs separated by newlines is also valid for import.
Its user friendly as you can edit with a text editor. It also supports services/browser extensions/etc. which export a plain text list of URLs.
An advantage of this format is it would make it trivial for people to migrate from Session Buddy. (Although Session Buddy also uses a JSON format) 
A while back I implemented a JS library for importing/exporting this format. It also supported the session buddy JSON format and possibly some other JSON variations seen by e.g. Chrome bookmarks export. I haven't released it anywhere, when I get a chance I may try to get it on Github.

**Comment by ReDEnergy** [comment]
Thanks for the feedback. Actually writing a Session Buddy importer it's a really good idea and I will try to do that for sure. It might help a few users that want to migrate from Session Buddy to Firefox.

Not really sure about the other way around since I'm not 100% sure that the storage is compatible between addons. I analyzed for a few minutes the extensions and I can see that there are several ways of exporting the sessions and I think I use those files to make the import without any problem.

P.S It's quite funny that Session Buddy is not released on Firefox because many users keep reminding me about it :). As I can see from the user based and ratings it's hugely popular, very simple to use and well designed.
It has that great session searching implementation and sessions can be saved under the same container, but other that that I don't really see a lot of other features that I don't have for the moment. At the same time, I do have lots of features that might not be considered as useful by everyone and/or I fail to present them in a good way, but It's really hard to try to offer all requested features and keep it as simple as Session Buddy looks. 

I do believe that what Session Buddy offers covers 99% of what users expect, but I don't plan to be a direct replacement for it. Session Sync started with a different idea in mind. From my point of view it's way way more flexible (more like an advanced bookmark management with session support) and I do plan to keep it that way.

That being said. I will write a basic importer for Session Buddy but I do have to make sure that user understand that Session Sync is not a direct replacement for it. It's an alternative that can do almost as much but at the same time so much more. Of course, I will try to cover all the missing features as soon as possible and I think I can do that.

Also... I really need to add support for 2 levels of bookmarks in order to save and display multi window session better.

**Comment by x11x** [comment]
Yes Session Sync is a great extension, for my purposes it does just about everything Session Buddy does. I'm very happy with Session Sync, I probably would not want to switch back to Firefox if it was not available.
The SB search works very well, it would be good to add more comprehensive search feature (e.g. by tab title) to Session Sync. Maybe you could use a 3rd party FTS library, like Lunr.js (I think that runs completely in memory, don't know if that would be suitable for this extension resource-wise).
Most of my requests would be the things already on your todo list (#40) -- you would definitely have more features than SB and a very decent replacement.
The drag and drop is good which SB doesn't have yet.
I think SB is in some ways a bit more intuitive and user friendly. Some features in Session Sync are a bit harder to find, like deleting and renaming tabs and sessions. But I like the overall design of SS.
One minor UI improvement off the top of my head: what if a ... icon appeared next to the items (sessions and tabs) when you hover, and that acted as a context menu with delete, rename, etc. Or you could even have some more common actions like delete appearing next to items on hover, to make it even quicker.
I really like the multiple item selection feature (#9) but I realise this might be a bigger change. Some things like this would make the UI more intuitive based on how other apps work.
Again, very good extension, thanks for your work on it.

**Comment by x11x** [comment]
Re the export format:
I think exporting in some format would be good, just for the sake of users having control of their data. You could choose a standard format for export, like the full Session Buddy format (with titles) or a JSON format like the one used in SB or Chrome bookmarks export.
Backup is important and I realize this is already possible using the browser's built-in bookmarks backup feature. But maybe this is not immediately obvious to some users. You could have a shortcut to that within Session Sync, or maybe just a backup button which displays a tip to the user telling them how to backup their bookmarks and explaining that sessions are also backed up when you do this.

**Comment by ReDEnergy** [comment]
Yes, I do plan to add that context-menu hover icon. Added a detailed entry in #40 just to be sure I won't miss it.
Regarding backup. Yes I really do need to offer something integrated into the addon. Most users will expect this and like you said many are not aware that the Bookmark manager provides such an option (sadly it exports all your bookmarks and you can't really export part of). Regarding exporting... this is a little bit tricky because there are several factors that affect this feature:
 - what's the purpose of the export
   - should it be compatible with some other extension import feature ? (do I really want to do that?)
   - is it only for backup purpose
 - user readable or a certain standard format
   - if I use standard bookmark backup format then it should be compatible with any browser but might be hard to read
 - then why not multiple formats ? well... 
   - it's quite hard to maintain
   - can create confusion 

Reality... From past experience I kinda know that once I provide an option I will be asked for the exact different one I didn't provide. Most probably I will start with the most simple one and add better options in the future.

Regarding of the friendly and intuitive UI ... it's a pain to fit so many things in that small hard limited 800 x 600 px panel. And even though I also provide the full page/window experience some things just can be done from a tab (Like adding the current tab to a session)

**Comment by x11x** [comment]
Thanks for your replies.
Yeah, the UI is great as it is and good use of the small panel. The improvements I've suggested are only minor and I guess the UI can evolve over time.
Re export format, I can see your points and its up to you.
I don't think there's any need to provide lots of different formats, only the easy or useful ones. E.g CSV or HTML could be good. I think compatibility with formats used by other tools is good but there's no need to go to extra trouble to be compatible.
Also, I would say you don't need to worry about losing users to other extensions by making export too easy, because if this extension is good, they won't want to change. I recognise the threat of Session Buddy if that ever comes to Firefox, but this extension is a good alternative and hopefully it will have an edge!

**Comment by ReDEnergy** [comment]
Not really afraid of the competition from Session Buddy on Firefox, the more alternatives the better. (Less negative reviews for me that I don't do what Session Buddy does :)).
I kinda more dislike the idea that I might provide a way out of Firefox for people. I want to do something for the Firefox community, because it had a huge impact over me for the last few years. Truth to be told, I do care about all this feedback for the addon because I know that most of it comes from a some wonderful people for whom Firefox is not just a browser and even though Chrome might be a little faster here and there they really care about it. 
So yeah... for to moment everything I plan to do it's mostly for the community (and myself) but less for the product overall, I do really care about it but it's built primary for the community. I mean, some of the features that are implemented I never even use them myself (for example the auto-saving feature, pinned tabs (yet to be added) and other things).
